File: 2025/R/solve_day7_part1.R
solve_day7_part1 <- function(input) {
  grid <- aoc_input_to_chr_matrix(input)

  nrows <- NROW(grid)
  ncols <- NCOL(grid)

  source <- which(grid == "S", arr.ind = TRUE)
#   splitters <- which(grid == "^", arr.ind = TRUE)

  row <- source[[1L]] + 1L
  beams <- vector("list", nrows)
  beams[[2L]] <- source[[2L]]
  split_count <- 0L

  while(row < nrows) {
    next_row <- row + 1L

    for (i in beams[[row]]) {
      if (grid[row, i] == "^") {
        split_count <- split_count + 1L
        beams[[next_row]] <- c(beams[[next_row]], i-1, i+1)
        next
      }

      if (!i %in% beams[[next_row]]) {
        beams[[next_row]] <- c(beams[[next_row]], i)
      }
    }

    row <- next_row
  }

  split_count
}

File: 2025/python/solve_day7_part1.py
def solve_day7_part1(text):
    nrows = len(text)
    ncols = len(text[0])

    source = text[0].index("S")
    beams = [[0] * ncols for _ in range(nrows)]
    beams[0][source] = 1

    split_count = 0

    for row in range(nrows-1):
        next_row = row + 1
        beam_cols = [i for i, val in enumerate(beams[row]) if val > 0]

        for col in beam_cols:
            if text[next_row][col] == "^":
                beams[next_row][col+1] += 1
                beams[next_row][col-1] += 1
                split_count += 1
                continue

            if beams[next_row][col] == 0:
                beams[next_row][col] = 1

    return(split_count)
